runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +943,200 @@ impl<'a> VMLogic<'a> {
     pub fn keccak512(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
         self.gas_counter.pay_base(keccak512_base)?;
         let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
-        self.gas_counter.pay_per_byte(keccak512_byte, value.len() as u64)?;
+        self.gas_counter.pay_per(keccak512_byte, value.len() as u64)?;
 
         use sha3::Digest;
 
         let value_hash = sha3::Keccak512::digest(&value);
         self.internal_write_register(register_id, value_hash.as_slice().to_vec())
     }
 
+    /// Hashes the given value using RIPEMD-160 and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    ///  Where `message_blocks` is `max(message_len / 8, 1)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ripemd160_base + ripemd160_block * message_blocks`
+    #[cfg(feature = "protocol_feature_evm")]
+    pub fn ripemd160(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
+        self.gas_counter.pay_base(ripemd160_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let message_blocks = std::cmp::max(value_len / 8, 1);
+
+        self.gas_counter.pay_per(ripemd160_block, message_blocks)?;
+
+        use ripemd160::Digest;
+
+        let value_hash = ripemd160::Ripemd160::digest(&value);
+        self.internal_write_register(register_id, value_hash.as_slice().to_vec())
+    }
+
+    /// Hashes the given message using BLAKE2b and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// If the message is too long and overflows, an error is returned with
+    /// `Blake2HashDataOverflow`.
+    ///
+    /// # Cost
+    ///
+    /// Where `message_blocks` is `max(message_len / 128, 1)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes +
+    /// message_blocks * blake2b_block + rounds * message_blocks`
+    #[cfg(feature = "protocol_feature_evm")]
+    pub fn blake2b(
+        &mut self,
+        rounds: u32,
+        state_len: u64,
+        state_ptr: u64,
+        message_len: u64,
+        message_ptr: u64,
+        t0: u64,
+        t1: u64,
+        f0: u64,
+        f1: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        use blake2::VarBlake2b;
+        use std::convert::TryFrom;
+
+        if state_len != 8 {
+            return Err(HostError::Blake2InvalidStateLength { length: state_len }.into());
+        }
```

Comment:
We probably should do this after paying the base cost? Or, rather, why do we even pass `state_len` if it *has* to be `8`?

I think that's the right pattern to use here:

https://github.com/near/nearcore/blob/f797a3b5585f1bec47bbe0348efd7e3728df2382/runtime/near-vm-logic/src/logic.rs#L1424